src/lib/util/VersionUpgradeManager.ts

```diff
@@ -0,0 +1,84 @@
+//   they should be sufficiently generic that providing them to other scripts isn't troublesome.
+// - You can assume that every previous upgrade script was successfully executed.
+// - Make sure that if your script fails, it fails cleanly. Roll back changes, and provide a clear error message.
+const upgradeScriptsByVersion: Map<string, VersionUpgradeScript> = new Map();
+upgradeScriptsByVersion.set('v2.7.0', (config: ConfigContent): Promise<void> => {
+	// v2.7.0 adds advanced target patterns and generally enhances the RetireJS integration, and some of the patterns
+	// that used to be in the config are now hardcoded into the engine. The config's targetPatterns array should be
+	// scoured of any such patterns. Other patterns should be permitted to stay, since they might be meaningful.
+	const retireJsConfig: EngineConfigContent = config.engines.find(e => e.name === ENGINE.RETIRE_JS);
+	if (retireJsConfig) {
+		const hardcodedPatterns: Set<string> = new Set(RetireJsEngine.getSimpleTargetPatterns());
+		retireJsConfig.targetPatterns = retireJsConfig.targetPatterns.filter(s => !hardcodedPatterns.has(s));
+	}
+	return Promise.resolve();
+});
+
+
+// ================ CLASSES =====================
+export class VersionUpgradeError extends Error {}
+
+export class VersionUpgradeManager {
+	private readonly currentVersion: string;
+	private upgradeScriptsByVersion: Map<string,VersionUpgradeScript>;
+
+	constructor() {
+		this.currentVersion = require('../../../package.json').version;
+		this.upgradeScriptsByVersion = upgradeScriptsByVersion;
+	}
+
+	public upgradeRequired(configVersion: string): boolean {
+		return !configVersion || (semver.lt(configVersion, this.currentVersion));
+	}
+
+	private getVersionsBetween(fromVersion: string, toVersion: string): string[] {
+		const orderedVersions = semver.sort([...this.upgradeScriptsByVersion.keys()]);
+		return orderedVersions.filter((v: string) => {
+			// The lower bound is EXCLUSIVE, and a null value is treated as below non-null value.
+			const afterFromVersion = !fromVersion || semver.lt(fromVersion, v);
+			// The upper bound is INCLUSIVE.
+			const atOrBeforeToVersion = semver.lte(v, toVersion);
+			return afterFromVersion && atOrBeforeToVersion;
+		});
+	}
+
+	private async upgrade(config: ConfigContent, fromVersion: string, toVersion: string): Promise<void> {
+		// We want to sequentially apply all of the upgrade scripts between the fromVersion and the toVersion, so we'll
+		// need to get all of those versions.
+		const versions: string[] = this.getVersionsBetween(fromVersion, toVersion);
+
+		// Handle each upgrade script in sequence.
+		for (const version of versions) {
+			try {
+				await this.upgradeScriptsByVersion.get(version)(config);
+			} catch (e) {
+				// If the upgrade script fails, prefix the error so it's clear where it came from, then rethrow it.
+				throw new VersionUpgradeError(`Upgrade script for ${version} failed: ${e.message || e}`);
```

Comment:
Most of the other code uses the pattern below, could this do the same? I think it would be helpful to the let the user know that they could remove the `${HOME}/.sfdx-scanner/Config.json` file if errors continues to occur.

```
throw SfdxError.create('@salesforce/sfdx-scanner', 'jreSetupManager', 'NoJavaHomeFound', []);
```

Reply:
I'll switch to that pattern. Also, good point, I'll add that to the error message.